File: src/main/java/com/wanderers/hotelier_webservices/mapper/AccommodationMapper.java
package com.wanderers.hotelier_webservices.mapper;

import com.wanderers.hotelier_webservices.rest.model.AccommodationRequestBody;
import com.wanderers.hotelier_webservices.rest.model.AccommodationResponseBody;
import com.wanderers.hotelier_webservices.rest.model.Location;
import com.wanderers.hotelier_webservices.server.dto.AccommodationDto;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Class is responsible for mapping {@link AccommodationDto} DTO between REST and Server layers
 */
@Component("accommodation_mapper")
public class AccommodationMapper {

    public AccommodationDto mapToDto(AccommodationRequestBody accommodationReq, String hotelierId) {
        final AccommodationDto accommodationDto = new AccommodationDto();

        accommodationDto.setHotelierId(hotelierId);
        accommodationDto.setName(accommodationReq.getName());
        accommodationDto.setRating(accommodationReq.getRating());
        accommodationDto.setCategory(accommodationReq.getCategory().getValue().toUpperCase());
        accommodationDto.setCity(accommodationReq.getLocation().getCity());
        accommodationDto.setState(accommodationReq.getLocation().getState());
        accommodationDto.setCountry(accommodationReq.getLocation().getCountry());
        accommodationDto.setZipCode(accommodationReq.getLocation().getZipCode());
        accommodationDto.setAddress(accommodationReq.getLocation().getAddress());
        accommodationDto.setImage(accommodationReq.getImage());
        accommodationDto.setReputation(accommodationReq.getReputation());
        accommodationDto.setPrice(accommodationReq.getPrice());
        accommodationDto.setAvailability(accommodationReq.getAvailability());

        return accommodationDto;
    }

    public AccommodationResponseBody mapToRestAccommodation(AccommodationDto accommodationDto) {
        final AccommodationResponseBody accommodationRes = new AccommodationResponseBody();
        final Location location = new Location();

        location.setCity(accommodationDto.getCity());
        location.setState(accommodationDto.getState());
        location.setCountry(accommodationDto.getCountry());
        location.setZipCode(accommodationDto.getZipCode());
        location.setAddress(accommodationDto.getAddress());

        accommodationRes.setId(accommodationDto.getId());
        accommodationRes.setName(accommodationDto.getName());
        accommodationRes.setRating(accommodationDto.getRating());
        accommodationRes.setCategory(AccommodationResponseBody.CategoryEnum.fromValue(accommodationDto.getCategory().toLowerCase()));
        accommodationRes.setLocation(location);
        accommodationRes.setImage(accommodationDto.getImage());
        accommodationRes.setReputation(accommodationDto.getReputation());
        accommodationRes.setReputationBadge(accommodationDto.getReputationBadge());
        accommodationRes.setPrice(accommodationDto.getPrice());
        accommodationRes.setAvailability(accommodationDto.getAvailability());
        accommodationRes.setName(accommodationDto.getName());
        accommodationRes.setName(accommodationDto.getName());

        return accommodationRes;
    }

    public List<AccommodationResponseBody> mapToRestAccommodations(List<AccommodationDto> accommodationDTOs) {
        return accommodationDTOs.stream().map(this::mapToRestAccommodation).collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/wanderers/hotelier_webservices/mapper/AccommodationMapper.kt
package com.wanderers.hotelier_webservices.mapper

import com.wanderers.hotelier_webservices.rest.model.AccommodationRequestBody
import com.wanderers.hotelier_webservices.rest.model.AccommodationResponseBody
import com.wanderers.hotelier_webservices.rest.model.Location
import com.wanderers.hotelier_webservices.server.dto.AccommodationDto
import org.springframework.stereotype.Component
import java.util.*
import java.util.stream.Collectors

/**
 * Class is responsible for mapping [AccommodationDto] DTO between REST and Server layers
 */
@Component("accommodation_mapper")
class AccommodationMapper {
    fun mapToDto(accommodationReq: AccommodationRequestBody, hotelierId: String?): AccommodationDto {
        val accommodationDto = AccommodationDto()
        accommodationDto.hotelierId(hotelierId)
        accommodationDto.name(accommodationReq.name)
        accommodationDto.rating(accommodationReq.rating)
        accommodationDto.category(accommodationReq.category.value.uppercase(Locale.getDefault()))
        accommodationDto.city(accommodationReq.location.city)
        accommodationDto.state(accommodationReq.location.state)
        accommodationDto.country(accommodationReq.location.country)
        accommodationDto.zipCode(accommodationReq.location.zipCode)
        accommodationDto.address(accommodationReq.location.address)
        accommodationDto.image(accommodationReq.image)
        accommodationDto.reputation(accommodationReq.reputation)
        accommodationDto.price(accommodationReq.price)
        accommodationDto.availability(accommodationReq.availability)
        return accommodationDto
    }

    fun mapToRestAccommodation(accommodationDto: AccommodationDto): AccommodationResponseBody {
        val accommodationRes = AccommodationResponseBody()
        val location = Location()
        location.city(accommodationDto.getCity())
        location.state(accommodationDto.getState())
        location.country(accommodationDto.getCountry())
        location.zipCode(accommodationDto.getZipCode())
        location.address(accommodationDto.getAddress())
        accommodationRes.id(accommodationDto.getId())
        accommodationRes.name(accommodationDto.getName())
        accommodationRes.rating(accommodationDto.getRating())
        accommodationRes.category(
            AccommodationResponseBody.CategoryEnum.fromValue(
                accommodationDto.getCategory().toLowerCase()
            )
        )
        accommodationRes.location(location)
        accommodationRes.image(accommodationDto.getImage())
        accommodationRes.reputation(accommodationDto.getReputation())
        accommodationRes.reputationBadge(accommodationDto.getReputationBadge())
        accommodationRes.price(accommodationDto.getPrice())
        accommodationRes.availability(accommodationDto.getAvailability())
        accommodationRes.name(accommodationDto.getName())
        accommodationRes.setName(accommodationDto.getName())
        return accommodationRes
    }

    fun mapToRestAccommodations(accommodationDTOs: List<AccommodationDto>): List<AccommodationResponseBody> {
        return accommodationDTOs.stream().map { accommodationDto: AccommodationDto ->
            mapToRestAccommodation(
                accommodationDto
            )
        }.collect(Collectors.toList())
    }
}

private operator fun String?.invoke(hotelierId: String?) {

}
